tools/sgcloudendpoints/tools.go

```diff
@@ -4,100 +4,159 @@ import (
+				"endpoints",
+				"services",
+				"deploy",
+				"--format",
+				"value(serviceConfig.id)",
+				"--project",
+				project,
+			},
+			configFiles...,
+		)...,
 	)
+	var output strings.Builder
+	cmd.Stdout = &output
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(output.String()), nil
+}
 
-	return sg.Output(sg.Command(
+// ValidateConfig validates the provided config files with Cloud Endpoints.
+func ValidateConfig(ctx context.Context, project string, configFiles ...string) error {
+	sg.Logger(ctx).Printf("validating config...")
+	cmd := sggcloud.Command(
 		ctx,
-		"gcloud",
-		"endpoints",
-		"configs",
-		"list",
-		"--service",
-		serviceName,
-		"--project",
-		gcpProject,
-		"--limit",
-		"1",
-		"--format",
-		"value(id)",
-	))
+		append(
+			[]string{
+				"endpoints",
+				"services",
+				"deploy",
+				"--project",
+				project,
+				"--validate-only",
+			},
+			configFiles...,
+		)...,
+	)
+	cmd.Stdout, cmd.Stderr = nil, nil // suppress noise
+	return cmd.Run()
 }
 
-func DockerImage(ctx context.Context, serviceName, gcpProject, gcpRegion string) string {
-	dir, err := os.MkdirTemp(os.TempDir(), "sgcloudendpoints")
+// BuildImage builds a container image with a baked-in Cloud Endpoints service configuration.
+func BuildImage(ctx context.Context, project, region, repo, service, configID string) (string, error) {
+	sg.Logger(ctx).Printf("building image...")
+	config, err := GetServiceConfig(ctx, service, configID)
+	if err != nil {
+		return "", err
+	}
+	buildRoot := sg.FromBuildDir("cloud-endpoints")
+	if err := os.MkdirAll(buildRoot, 0o700); err != nil {
+		return "", err
+	}
+	buildDir, err := os.MkdirTemp(buildRoot, "docker")
 	if err != nil {
-		panic(err)
+		return "", err
+	}
+	defer func() {
+		_ = os.RemoveAll(buildDir)
+	}()
+	if err := os.WriteFile(filepath.Join(buildDir, "Dockerfile"), dockerfile, 0o600); err != nil {
+		return "", err
+	}
+	if err := os.WriteFile(filepath.Join(buildDir, "service.json"), []byte(config), 0o600); err != nil {
+		return "", err
 	}
-	configID := ConfigID(ctx, serviceName, gcpProject)
-	sg.Logger(ctx).Printf(
-		"building image for %s with configID %s...",
-		serviceName,
+	image := fmt.Sprintf(
+		"%s-docker.pkg.dev/%s/%s/endpoints-runtime-serverless:%s-%s-%s",
+		region,
+		project,
+		repo,
+		espVersion,
+		service,
 		configID,
 	)
-	req, err := http.NewRequestWithContext(
+	if err := sgdocker.Command(
+		ctx, "build", "-t", image, "--build-arg", "esp_version="+espVersion, buildDir,
+	).Run(); err != nil {
+		return "", err
+	}
+	return image, nil
+}
+
+// GetServiceConfig fetches a full service config from Cloud Endpoints.
+func GetServiceConfig(ctx context.Context, service, configID string) (string, error) {
+	var accessTokenOutput strings.Builder
+	cmd := sggcloud.Command(ctx, "auth", "print-access-token")
+	cmd.Stdout = &accessTokenOutput
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+	request, err := http.NewRequestWithContext(
```

Comment:
Isn't this simpler to do with gcloud directly. Something like

```
gcloud endpoints configs describe --service=my-service <Config ID>
```

Reply:
Hmm! That sounds right! Let me try it

Reply:
@alethenorio Updated

Reply:
@alethenorio I tested it out and it seems that the output is not 100% equivalent (the service refuses to start when this config is embedded into the container).

I think it has to do with the `view=FULL` parameter which is not exposed via the CLI.

Reply:
Ah that's too bad. Fetching through http works just fine as well

Reply:
Let's go with the HTTP fetch for now? We got it from here, so there's probably intent behind it: https://github.com/GoogleCloudPlatform/esp-v2/blob/master/docker/serverless/gcloud_build_image

Reply:
Yes let's do that